File: bindings/elixir/lib/rockbox/metadata.ex
defmodule Rockbox.Metadata do
  @moduledoc "Audio file metadata / tag parsing."

  @doc """
  Parse the metadata of the audio file at `path`.

  Returns `{:ok, map}` with atom keys (`:codec`, `:title`, `:duration_ms`,
  `:sample_rate`, a nested `:replaygain` map, …) or `{:error, reason}`.
  """
  @spec read(Path.t()) :: {:ok, map()} | {:error, term()}
  def read(path) do
    case :rockbox_ffi_nif.meta_read_json(to_bin(path)) do
      nil -> {:error, :parse_failed}
      json when is_binary(json) -> {:ok, Rockbox.decode_json(json)}
    end
  end

  @doc "Same as `read/1` but raises on error."
  @spec read!(Path.t()) :: map()
  def read!(path) do
    case read(path) do
      {:ok, meta} -> meta
      {:error, reason} -> raise "rockbox metadata read failed: #{inspect(reason)}"
    end
  end

  @doc """
  Guess the codec label (e.g. `"FLAC"`) from a filename's extension without
  opening the file. Returns `nil` for an unknown extension.
  """
  @spec probe(String.t()) :: String.t() | nil
  def probe(filename) do
    case :rockbox_ffi_nif.meta_probe(to_bin(filename)) do
      nil -> nil
      label when is_binary(label) -> label
    end
  end

  defp to_bin(p), do: IO.iodata_to_binary([p])
end

File: bindings/swift/Sources/RockboxFFI/Metadata.swift
import Foundation

/// Audio file metadata parsing.
public enum Metadata {
    /// Parse the metadata of the audio file at `path`.
    ///
    /// Returns a dictionary with tag strings, `duration_ms`, `bitrate`,
    /// `sample_rate`, a nested `replaygain` dictionary (including the raw
    /// Q7.24 `raw_*` fields), and optional `album_art` / `cuesheet`
    /// locations. Throws on failure.
    public static func read(_ path: String) throws -> [String: Any] {
        let lib = Lib.shared
        let raw = path.withCString { lib.metaReadJson($0) }
        guard let json = lib.takeString(raw) else {
            throw RockboxError.nullReturn("rb_meta_read_json(\(path))")
        }
        return try parseObject(json, context: "metadata")
    }

    /// Guess the codec label (e.g. "FLAC") from a filename's extension
    /// without opening the file. Returns nil for an unknown extension.
    public static func probe(_ filename: String) -> String? {
        let lib = Lib.shared
        let raw = filename.withCString { lib.metaProbe($0) }
        return lib.takeString(raw)
    }
}

/// Parse a JSON object string into a dictionary.
func parseObject(_ json: String, context: String) throws -> [String: Any] {
    guard let data = json.data(using: .utf8),
          let obj = try JSONSerialization.jsonObject(with: data) as? [String: Any]
    else {
        throw RockboxError.invalidInput("could not parse \(context) JSON")
    }
    return obj
}
